extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+  )
+  {
+    this.consumerProps = consumerProps;
+    this.pullBatchSize = pullBatchSize;
+    for (Map.Entry<Object, Object> configItem : this.consumerProps.entrySet()) {
+      System.setProperty(configItem.getKey().toString(), configItem.getValue().toString());
+    }
+    this.consumerGroup = consumerGroup;
+    this.feed = feed;
+  }
+
+  /**
+   * Check if there are locally pending messages to consume.
+   *
+   * @return true if there are some; false otherwise.
+   */
+  private boolean hasMessagesPending()
+  {
+
+    for (ConcurrentHashMap.Entry<MessageQueue, ConcurrentSkipListSet<MessageExt>> entry : messageQueueTreeSetMap.entrySet()) {
+      if (!entry.getValue().isEmpty()) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public Firehose connect(ByteBufferInputRowParser byteBufferInputRowParser) throws IOException, ParseException
+  {
+
+    Set<String> newDimExclus = Sets.union(
+        byteBufferInputRowParser.getParseSpec().getDimensionsSpec().getDimensionExclusions(),
+        Sets.newHashSet("feed")
+    );
+
+    final ByteBufferInputRowParser theParser = byteBufferInputRowParser.withParseSpec(
+        byteBufferInputRowParser.getParseSpec()
+                                .withDimensionsSpec(
+                                    byteBufferInputRowParser.getParseSpec()
+                                                            .getDimensionsSpec()
+                                                            .withDimensionExclusions(
+                                                                newDimExclus
+                                                            )
+                                )
+    );
+
+    /**
+     * Topic-Queue mapping.
+     */
+    final ConcurrentHashMap<String, Set<MessageQueue>> topicQueueMap;
+
+    /**
+     * Default Pull-style client for RocketMQ.
+     */
+    final DefaultMQPullConsumer defaultMQPullConsumer;
+    final DruidPullMessageService pullMessageService;
+
+    messageQueueTreeSetMap.clear();
+    windows.clear();
+
+    try {
+      defaultMQPullConsumer = new DefaultMQPullConsumer(this.consumerGroup);
+      defaultMQPullConsumer.setMessageModel(MessageModel.CLUSTERING);
+      topicQueueMap = new ConcurrentHashMap<>();
+
+      pullMessageService = new DruidPullMessageService(defaultMQPullConsumer);
+      for (String topic : feed) {
+        Validators.checkTopic(topic);
+        topicQueueMap.put(topic, defaultMQPullConsumer.fetchSubscribeMessageQueues(topic));
+      }
+      DruidMessageQueueListener druidMessageQueueListener =
+          new DruidMessageQueueListener(Sets.newHashSet(feed), topicQueueMap, defaultMQPullConsumer);
+      defaultMQPullConsumer.setMessageQueueListener(druidMessageQueueListener);
+      defaultMQPullConsumer.start();
+      pullMessageService.start();
+    }
+    catch (MQClientException e) {
+      LOGGER.error("Failed to start DefaultMQPullConsumer", e);
```

Comment:
OK, will remove the logging and assuming the caller to handle/log the exception.